src/openfl/text/_internal/ShapeCache.hx

```diff
@@ -0,0 +1,109 @@
+package openfl.text._internal;
+import haxe.ds.IntMap;
+import haxe.ds.StringMap;
+
+#if !openfl_debug
+@:fileXml('tags="haxe,release"')
+@:noDebug
+#end
+
+@:access(openfl.text.TextFormat)
+@SuppressWarnings("checkstyle:FieldDocComment")
+class ShapeCache
+{
+	private var __shortWordMap:StringMap<StringMap<Array<GlyphPosition>>>;
+	private var __longWordMap:StringMap<IntMap<CacheMeasurement>>;
+
+	public function new()
+	{
+		__shortWordMap = new StringMap();
+		__longWordMap = new StringMap();
+	}
+
+	private static function hashFunction(key:String):Int
+	{
+		var hash = 0, i, chr;
+		for (i in 0...key.length)
+		{
+			chr = key.charCodeAt(i);
+			hash = ((hash << 5) - hash) + chr;
+			hash |= 0;
+		}
+		return hash;
+	}
+
+	public function cache(formatRange:TextFormatRange, getPositions:#if (js && html5) ()->, wordKey:String = null #else TextLayout #end):Dynamic
+	{
+		var formatKey:String = formatRange.format.__cacheKey;
+		#if (!js && !html5)
+		var wordKey:String = getPositions.text;
+		#end
+		if (wordKey.length > 15)
+		{
+			return __cacheLongWord(wordKey, formatKey, getPositions);
+		}
+		else {
+			return __cacheShortWord(wordKey, formatKey, getPositions);
+		}
+		
+	}
+
+	private function __cacheShortWord(wordKey:String, formatKey:String, getPositions:#if (js && html5) ()->):Array<Float #else TextLayout):Array<GlyphPosition #end>
+	{
+		if (__shortWordMap.exists(formatKey))
+		{
+			var formatMap = __shortWordMap.get(formatKey);
+			if (formatMap.exists(wordKey))
+			{
+				return formatMap.get(wordKey);
+			}
+			else
+			{
+				formatMap.set(wordKey, #if (js && html5) getPositions() #else getPositions.positions #end);
+			}
+		}
+		else
+		{
+			var formatMap = new StringMap();
+			formatMap.set(wordKey, #if (js && html5) getPositions() #else getPositions.positions #end);
```

Comment:
It is possible to store js positions as GlyphPosition objects instead of floats. Doing so here would unify the API everywhere else I imagine

Reply:
There's a few things that would make this more simple. This was the hardest problem I faced nearing completion....how to avoid having a api separation between html5 and native. Instead of doing that to accommodate for the separation of getPositions in TextEngine, I tried to architect it so that we could unify some of the things. GlyphPositions would be more expensive to create, but would make code more readable. I'm more interested in consolidating the getPositions code in TE.